### PR TITLE
custom substitutions can have phrases and substitutions

### PR DESCRIPTION
**Pull request by UnorderedSigh**
**Feature:** This PR allows phrases and substitutions in custom substitutions.

~~Related issue #6981~~ - not so related, now that I read it more carefully

## Feature Details
Allows phrases and substitutions in custom substitutions. Avoids infinite recursion using an efficient algorithm.

```ruby
mission whatever
    ...
    substitution
       <you> "${you}"
       <band> "${band}"
       <speaker> "<band> drummer"
       <event> "${band event}"
    on offer
        conversation
            `"Hello there," says the <speaker>, "you must be <you>. Go fly to <stopovers>! Then return to <origin> for a our <event>!"`
            choice
                `    Yup.`
                   accept
                `    Nope.`
                   decline
    on complete
        conversation
            `You land on <origin> after your tiring trip to <stopovers> and back, and locate the <band>'s <event>. Shortly after, they give you <payment>.`

phrase you
    word
        "Captain <last>"
        "<first> <last>"

phrase "band event"
    word
        "concert"
        "show"
        "party"
```

## UI Screenshots
A composite of five screenshots. There is reoccurring text throughout the conversation, copied from the substitutions used throughout the mission. Sender, recipient, package, and the mob that mobs you, are all consistent. Notice the refusal has three infinitely recursive substitution patterns, all of which are caught. The first offending tag is not expanded.

phrase-substitution-test

## Usage Examples
Randomly-generated missions: you can have a character or organization that shows up throughout the mission with the same name.

## Testing Done
1. Put this in data/[substest.txt](https://github.com/endless-sky/endless-sky/files/10609624/substest.txt)
2. Land on any planet with a spaceport.
3. Visit the spaceport.
4. Accept the mission or decline it.
5. If you accept, fly to the stopover.
6. Fly to the final destination to complete the mission. Notice that the same text from the origin and stopover is re-used here.

### Automated Tests Added
N/A

## Performance Impact
Marginally faster.

The Format::Replace now uses a map::find instead of a loop over all keys. I expect the loop over all keys was done to eliminate the cost of allocating a string each time. I avoid the cost by re-using the same string buffer to store each key via string::assign.

Recursive expansion of the map of all substitutions is done in Format::Expand, which uses a very fast algorithm to detect infinite recursion. The example txt I gave has three infinite recursive substitutions in the text for the player refusing the mission.

### CODE REVIEW

**Comment by bene-dictator** [comment]
This would be really neat for a few jobs I'm doing right now.

File: source/Mission.cpp [hunk 1]
```diff
@@ -950,6 +951,10 @@ string Mission::BlockedMessage(const PlayerInfo &player)
 		out << "no additional space";
 	subs["<capacity>"] = out.str();
 
+	for(auto &keyValue : subs)
```
Comment (RisingLeaf) [inline source/Mission.cpp:954]:
```suggestion
	for(const auto &keyValue : subs)
```

coders are stupid sometimes, so dont give them the opportunity to be XD

File: source/Mission.cpp [hunk 2]
```diff
@@ -1378,6 +1383,11 @@ Mission Mission::Instantiate(const PlayerInfo &player, const shared_ptr<Ship> &b
 		subs["<waypoints>"] = systems;
 	}
 
+	// Done making subs, so expand the phrases and recursively substitute.
+	for(auto &keyValue : subs)
```
Comment (RisingLeaf) [inline source/Mission.cpp:1387]:
```suggestion
	for(const auto &keyValue : subs)
```
same here.

or am I overlooking something?

File: source/text/Format.cpp [hunk 3]
```diff
@@ -258,38 +346,30 @@ string Format::Replace(const string &source, const map<string, string> &keys)
...
-	result.append(source, start, source.length() - start);
-	return result;
+void Format::Expand(map<string, string> &keys)
+{
+	map<string, string> newKeys;
+	unordered_set<string> keysBeingExpanded;
+	for(auto it = keys.begin(); it != keys.end(); ++it)
+		if(newKeys.find(it->first) == newKeys.end())
+			ExpandInto(it->first, it->second, keys, newKeys, keysBeingExpanded);
+	keys.swap(newKeys);
```
Comment (RisingLeaf) [inline source/text/Format.cpp:372]: I am confused(as always).
what do we need keysBeingExpanded for?

**Review by RisingLeaf — CHANGES_REQUESTED** [review]
quick review

File: source/text/Format.cpp (hunk 3 above)
Comment (RisingLeaf) [reply source/text/Format.cpp:372]: oh wait i get it now nevermind

Comment (UnorderedSigh) [reply source/text/Format.cpp:372]: That's how it detects a cycle in the graph.

**Comment by UnorderedSigh** [comment]
Last tested and confirmed to work on Sun, 19 Feb 2023 03:53:48 UTC

**Comment by UnorderedSigh** [comment]
Last tested and confirmed to work at Sat, 04 Mar 2023 17:43:12 UTC.

**Comment by UnorderedSigh** [comment]
Last tested and known to work Sat, 25 Mar 2023 22:25:43.

**Comment by UnorderedSigh** [comment]
Last tested and known to work Sun, 16 Apr 2023 21:52:15 UTC

**Comment by UnorderedSigh** [comment]
Last tested and known to work Fri, 12 May 2023 02:18:17 UTC

**Comment by UnorderedSigh** [comment]
Last tested and known to work Mon, 29 May 2023 16:50:43 UTC

**Comment by UnorderedSigh** [comment]
Last tested and known to work Wed, 23 Aug 2023 03:40:32 UTC

**Comment by UnorderedSigh** [comment]
Last tested and known to work Wed, 22 Nov 2023 23:11:51 UTC

File: source/text/Format.cpp [hunk 4]
```diff
@@ -44,6 +47,91 @@ namespace {
...
+				break;
+
+			++right;
+			size_t length = right - left;
+			key.assign(source, left, length);
+			const string *subs = SubstitutionFor(key);
+			if(subs)
+			{
+				ProcessText(source, start, left - start);
+				ProcessText(*subs, 0, string::npos);
```
Comment (warp-core) [inline source/text/Format.cpp:74]: If we're using a abbreviation, use the singular `sub`.
`subs` could just be the first four characters of `substitution` but it looks more like the plural of `sub`, but this is only a single substitution.
```suggestion
			const string *sub = SubstitutionFor(key);
			if(sub)
			{
				ProcessText(source, start, left - start);
				ProcessText(*sub, 0, string::npos);
```

File: source/text/Format.cpp [hunk 5]
```diff
@@ -318,38 +406,30 @@ string Format::Replace(const string &source, const map<string, string> &keys)
...
-	}
 
-	result.append(source, start, source.length() - start);
-	return result;
+void Format::Expand(map<string, string> &keys)
+{
+	map<string, string> newKeys;
+	unordered_set<string> keysBeingExpanded;
+	for(auto it = keys.begin(); it != keys.end(); ++it)
+		if(newKeys.find(it->first) == newKeys.end())
```
Comment (warp-core) [inline source/text/Format.cpp:430]:
```suggestion
		if(!newKeys.count(it->first))
```

File: source/text/Format.cpp [hunk 6]
```diff
@@ -44,6 +47,91 @@ namespace {
 		reverse(result.begin(), result.end());
 	}
 
+	void ScanForSubstitutions(const string &source,
+			function<void(const string &, size_t, size_t)> ProcessText,
+			function<const string *(const string &)> SubstitutionFor)
```
Comment (warp-core) [inline source/text/Format.cpp:52]: I see this method called in two places.
The functions passed as `ProcessText` seem basically identical.

From `ExpandInto`:
```c++
		auto ProcessText = [&](const string &from, size_t start, size_t length)
		{
			newValue.append(from, start, length);
		};
```
In `Format::Expand`:
```c++
	auto ProcessText = [&](const string &from, size_t start, size_t length)
	{
		result.append(from, start, length);
	};
```

So, effectively, every time `ScanForSubstitutions` calls `ProcessText`, it is appending to some string.
Why not just have a local string in `ScanForSubstitutions` that gets appended to and then return that at the end?

File: source/text/Format.cpp [hunk 7]
```diff
@@ -44,6 +47,91 @@ namespace {
...
+			if(hasResult != result.end())
+				// Already finished this one.
+				return &hasResult->second;
+			if(keysBeingExpanded.find(request) != keysBeingExpanded.end())
+				// Refuse to traverse a cycle in the graph.
+				return nullptr;
+			auto hasSource = source.find(request);
+			if(hasSource == source.end())
+				// Undefined key.
+				return nullptr;
```
[the remark this answers is not in the document]
Comment (UnorderedSigh) [reply source/text/Format.cpp:121]: No. Count counts the number of keys. For a map, this is a trivial operation. For anything else, it'll become O(N). Using count in place of find is a bad practice.

File: source/text/Format.cpp (hunk 5 above)
Comment (UnorderedSigh) [reply source/text/Format.cpp:430]: No. Count counts the number of keys. For a map, this is a trivial operation. For anything else, it'll become O(N). Using count in place of find is a bad practice.

File: source/text/Format.cpp (hunk 4 above)
Comment (UnorderedSigh) [reply source/text/Format.cpp]: I applied this suggestion  manually.

File: source/text/Format.cpp (hunk 6 above)
Comment (UnorderedSigh) [reply source/text/Format.cpp]: Good catch. I replaced ProcessText with a string return value. Now that the function does more than scan, I renamed it to StringSubstituter

**Comment by UnorderedSigh** [comment]
Last tested and known to work Fri, 31 May 2024 21:01:21 UTC

File: source/text/Format.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (UnorderedSigh) [reply source/text/Format.cpp:206]: I misread this. You're also asking me to move the comments. I'll do that momentarily...

**Review by tibetiroka — APPROVED** [review]

**Review by TheGiraffe3 — APPROVED** [review]
This would be pretty nice to have.

**Pull request merged**